server/src/main/java/org/elasticsearch/index/mapper/AbstractGeometryFieldMapper.java

```diff
@@ -69,6 +69,7 @@
         public static final Explicit<Boolean> IGNORE_Z_VALUE = new Explicit<>(true, false);
     }
 
+    protected static List<ParserHandler> PARSER_EXTENSIONS = new ArrayList<>();
```

Comment:
I do not think the parser-extension part of this PR is needed. I think that the lack of `doc_values` parsing support in the field mapper is a bug in our field-mapper infrastructure, not a feature that needs to be extended by a plugin.

I would want to see the equivalent to this PR: https://github.com/elastic/elasticsearch/pull/47519 getting merged into master, so that this PR can focus on the Indexer extension points. what do you think?

Reply:
That would work except the builder is not overridden in xpack. It's the responsibility of the parser-extension (`parser.config`) to throw "doc values not supported" exceptions at mapping time as opposed to throwing them at index time. So when geo_shape doc values are moved to xpack we will want OSS behavior to remain the same (throwing doc values not supported exceptions) and xpack to implement them.

Reply:
I see, so these handlers must be passed in, but can we re-use the data-handler for this instead of creating a generic parsing abstraction? Would it be better to limit scope to just changing how doc-values are parsed?

Reply:
Yeah I originally had it that way but changed it because the data handler logic is specific to `GeoShapeFieldMapper` only (since index logic for both `ShapeFieldMapper` and `LegacyGeoShapeFieldMapper` is untouched). But parsing logic needs to be handled by both `GeoShapeFieldMapper` and `ShapeFieldMapper` so it was elevated to the `AbstractGeometryFieldMapper` in order to share the implementation (and not duplicate in `ShapeFieldMapper`). 

This is just the start to the "tangled mess" I was referring to in today's meeting. It gets more hairy when projections (`Geo3DPoint` `XYPoint` `LatLonPoint`) are brought in the fray.

Reply:
Would it work if there were dummy data-handlers/indexers unique for ShapeFieldMapper and LegacyGeoShapeFieldMapper?

regardless of how the extension point is implemented, I feel like it is important that all these field mappers understand how to parse and proclaim their lack of support for doc-values.

Do you see any reason why I should not add support for parsing `doc_values` to server's implementation?

Reply:
I think the implementation is mostly based on preference? I don't see a reason to not add support for parsing doc_values in server's implementation (and throwing a `MapperParsingException` in OSS deployments). I ultimately didn't go this route because I liked the existing behavior where unimplemented parameters are handled further upstream.

Reply:
Like @talevy I have a preference for not making parsing pluggable. We might need to reconsider in the future with projections, but I'd like to explore the path that requires the least amount of new abstractions for now?

Reply:
I opened #53351 to support parsing `doc_values`. There are two places I left TODO comments where I'd like to be sure this data-handler branch can properly hook into things.

let me know what you think!

Reply:
:+1:  I wired up the data handler. Next will make the change to not register the default data handler unless no plugins are installed.